datafusion/expr/src/expr_schema.rs

```diff
@@ -134,9 +134,25 @@ impl ExprSchemable for Expr {
                             )
                         })?;
 
+                        // perform additional function arguments validation (due to limited
+                        // expressiveness of `TypeSignature`), then infer return type
                         fun.return_type(&arg_data_types)
                     }
                     ScalarFunctionDefinition::UDF(fun) => {
+                        // verify that function is invoked with correct number and type of arguments as defined in `TypeSignature`
+                        data_types(&arg_data_types, fun.signature()).map_err(|_| {
+                            plan_datafusion_err!(
+                                "{}",
+                                utils::generate_signature_error_msg(
+                                    fun.name(),
+                                    fun.signature().clone(),
+                                    &arg_data_types,
+                                )
+                            )
+                        })?;
```

Comment:
Thanks for this! Doing it during the planning stage eliminates so much duplicate code:)

FYI @jonahgao, this addresses the TODO we talked about in this [comment](https://github.com/apache/arrow-datafusion/pull/9297#issuecomment-1958566729)

Reply:
Nice! 👍